File: src/main/java/waterpower/integration/minetweaker/Machines.java
/**
 * Copyright (c) Huang Yuhui, 2014
 * 
 * "WaterCraft" is distributed under the terms of the Minecraft Mod Public
 * License 1.0, or MMPL. Please check the contents of the license located in
 * http://www.mod-buildcraft.com/MMPL-1.0.txt
 */
package waterpower.integration.minetweaker;

import minetweaker.MineTweakerAPI;
import minetweaker.api.item.IItemStack;
import minetweaker.api.minecraft.MineTweakerMC;
import net.minecraft.item.ItemStack;

import waterpower.common.recipe.IRecipeManager;
import waterpower.common.recipe.MyRecipeOutput;
import waterpower.common.recipe.MyRecipes;

import stanhebben.zenscript.annotations.NotNull;
import stanhebben.zenscript.annotations.ZenClass;
import stanhebben.zenscript.annotations.ZenMethod;

@ZenClass("mods.waterpower.Machines")
public class Machines {

    private Machines() {
    }

    @ZenMethod
    public static void addCrusherRecipe(@NotNull IItemStack input, @NotNull IItemStack output) {
        MineTweakerAPI.apply(new AddRecipeAction("WaterPower Crusher", MyRecipes.macerator, MineTweakerMC.getItemStack(input), MineTweakerMC
                .getItemStack(output)));
    }

    @ZenMethod
    public static void addLatheRecipe(@NotNull IItemStack input, @NotNull IItemStack output) {
        MineTweakerAPI.apply(new AddRecipeAction("WaterPower Lathe", MyRecipes.lathe, MineTweakerMC.getItemStack(input), MineTweakerMC.getItemStack(output)));
    }

    @ZenMethod
    public static void addCutterRecipe(@NotNull IItemStack input, @NotNull IItemStack output) {
        MineTweakerAPI.apply(new AddRecipeAction("WaterPower Cutter", MyRecipes.cutter, MineTweakerMC.getItemStack(input), MineTweakerMC.getItemStack(output)));
    }

    @ZenMethod
    public static void addCompressorRecipe(@NotNull IItemStack input, @NotNull IItemStack output) {
        MineTweakerAPI.apply(new AddRecipeAction("WaterPower Compressor", MyRecipes.compressor, MineTweakerMC.getItemStack(input), MineTweakerMC
                .getItemStack(output)));
    }

    @ZenMethod
    public static void addCentrifugeRecipe(@NotNull IItemStack input, @NotNull IItemStack output) {
        MineTweakerAPI.apply(new AddRecipeAction("WaterPower Centrifuge", MyRecipes.centrifuge, MineTweakerMC.getItemStack(input), MineTweakerMC
                .getItemStack(output)));
    }

    @ZenMethod
    public static void addAdvancedCompressorRecipe(@NotNull IItemStack input, @NotNull IItemStack output) {
        MineTweakerAPI.apply(new AddRecipeAction("WaterPower Advanced Compressor", MyRecipes.implosion, MineTweakerMC.getItemStack(input), MineTweakerMC
                .getItemStack(output)));
    }

    @ZenMethod
    public static void addSawmillRecipe(@NotNull IItemStack input, @NotNull IItemStack output) {
        MineTweakerAPI
                .apply(new AddRecipeAction("WaterPower Sawmill", MyRecipes.sawmill, MineTweakerMC.getItemStack(input), MineTweakerMC.getItemStack(output)));
    }

    @ZenMethod
    public static void removeCrusherRecipe(@NotNull IItemStack input) {
        MineTweakerAPI.apply(new RemoveRecipeAction("WaterPower Crusher", MyRecipes.macerator, MineTweakerMC.getItemStack(input)));
    }

    @ZenMethod
    public static void removeLatheRecipe(@NotNull IItemStack input) {
        MineTweakerAPI.apply(new RemoveRecipeAction("WaterPower Lathe", MyRecipes.lathe, MineTweakerMC.getItemStack(input)));
    }

    @ZenMethod
    public static void removeCutterRecipe(@NotNull IItemStack input) {
        MineTweakerAPI.apply(new RemoveRecipeAction("WaterPower Cutter", MyRecipes.cutter, MineTweakerMC.getItemStack(input)));
    }

    @ZenMethod
    public static void removeCompressorRecipe(@NotNull IItemStack input) {
        MineTweakerAPI.apply(new RemoveRecipeAction("WaterPower Compressor", MyRecipes.compressor, MineTweakerMC.getItemStack(input)));
    }

    @ZenMethod
    public static void removeCentrifugeRecipe(@NotNull IItemStack input) {
        MineTweakerAPI.apply(new RemoveRecipeAction("WaterPower Centrifuge", MyRecipes.centrifuge, MineTweakerMC.getItemStack(input)));
    }

    @ZenMethod
    public static void removeAdvancedCompressorRecipe(@NotNull IItemStack input) {
        MineTweakerAPI.apply(new RemoveRecipeAction("WaterPower Advanced Compressor", MyRecipes.implosion, MineTweakerMC.getItemStack(input)));
    }

    @ZenMethod
    public static void removeSawmillRecipe(@NotNull IItemStack input, @NotNull IItemStack output) {
        MineTweakerAPI.apply(new RemoveRecipeAction("WaterPower Sawmill", MyRecipes.sawmill, MineTweakerMC.getItemStack(input)));
    }

    protected static IItemStack getMachineOutput(IItemStack input, IRecipeManager rm) {
        MyRecipeOutput output = rm.getOutput(MineTweakerMC.getItemStack(input), false);
        if ((output == null) || (output.items.isEmpty()))
            return null;
        return MineTweakerMC.getIItemStack((ItemStack) output.items.get(0));
    }

    @ZenMethod
    public static IItemStack getCrusherOutput(IItemStack input) {
        return getMachineOutput(input, MyRecipes.macerator);
    }

    @ZenMethod
    public static IItemStack getAdvancedCompressorOutput(IItemStack input) {
        return getMachineOutput(input, MyRecipes.implosion);
    }

    @ZenMethod
    public static IItemStack getCentrifugeOutput(IItemStack input) {
        return getMachineOutput(input, MyRecipes.centrifuge);
    }

    @ZenMethod
    public static IItemStack getCompressorOutput(IItemStack input) {
        return getMachineOutput(input, MyRecipes.compressor);
    }

    @ZenMethod
    public static IItemStack getCutterOutput(IItemStack input) {
        return getMachineOutput(input, MyRecipes.cutter);
    }

    @ZenMethod
    public static IItemStack getLatheOutput(IItemStack input) {
        return getMachineOutput(input, MyRecipes.lathe);
    }

    @ZenMethod
    public static IItemStack getSawmillOutput(IItemStack input) {
        return getMachineOutput(input, MyRecipes.sawmill);
    }

}

File: src/main/kotlin/waterpower/integration/crafttweaker/Machines.kt
/**
 * Copyright (c) Huang Yuhui, 2017
 *
 * "WaterPower" is distributed under the terms of the Minecraft Mod Public
 * License 1.0, or MMPL. Please check the contents of the license located in
 * http://www.mod-buildcraft.com/MMPL-1.0.txt
 */
package waterpower.integration.crafttweaker

import minetweaker.MineTweakerAPI
import minetweaker.api.item.IItemStack
import minetweaker.api.minecraft.MineTweakerMC
import stanhebben.zenscript.annotations.NotNull
import stanhebben.zenscript.annotations.ZenClass
import stanhebben.zenscript.annotations.ZenMethod
import waterpower.common.recipe.IRecipeManager
import waterpower.common.recipe.RecipeManagers


@ZenClass("mods.waterpower.Machines")
object Machines {

    @ZenMethod
    fun addCrusherRecipe(@NotNull input: IItemStack, @NotNull output: IItemStack) {
        MineTweakerAPI.apply(AddRecipeAction("WaterPower Crusher", RecipeManagers.crusher, MineTweakerMC.getItemStack(input), MineTweakerMC
                .getItemStack(output)))
    }

    @ZenMethod
    fun addLatheRecipe(@NotNull input: IItemStack, @NotNull output: IItemStack) {
        MineTweakerAPI.apply(AddRecipeAction("WaterPower Lathe", RecipeManagers.lathe, MineTweakerMC.getItemStack(input), MineTweakerMC.getItemStack(output)))
    }

    @ZenMethod
    fun addCutterRecipe(@NotNull input: IItemStack, @NotNull output: IItemStack) {
        MineTweakerAPI.apply(AddRecipeAction("WaterPower Cutter", RecipeManagers.cutter, MineTweakerMC.getItemStack(input), MineTweakerMC.getItemStack(output)))
    }

    @ZenMethod
    fun addCompressorRecipe(@NotNull input: IItemStack, @NotNull output: IItemStack) {
        MineTweakerAPI.apply(AddRecipeAction("WaterPower Compressor", RecipeManagers.compressor, MineTweakerMC.getItemStack(input), MineTweakerMC
                .getItemStack(output)))
    }

    @ZenMethod
    fun addCentrifugeRecipe(@NotNull input: IItemStack, @NotNull output: IItemStack) {
        MineTweakerAPI.apply(AddRecipeAction("WaterPower Centrifuge", RecipeManagers.centrifuge, MineTweakerMC.getItemStack(input), MineTweakerMC
                .getItemStack(output)))
    }

    @ZenMethod
    fun addAdvancedCompressorRecipe(@NotNull input: IItemStack, @NotNull output: IItemStack) {
        MineTweakerAPI.apply(AddRecipeAction("WaterPower Advanced Compressor", RecipeManagers.advCompressor, MineTweakerMC.getItemStack(input), MineTweakerMC
                .getItemStack(output)))
    }

    @ZenMethod
    fun addSawmillRecipe(@NotNull input: IItemStack, @NotNull output: IItemStack) {
        MineTweakerAPI
                .apply(AddRecipeAction("WaterPower Sawmill", RecipeManagers.sawmill, MineTweakerMC.getItemStack(input), MineTweakerMC.getItemStack(output)))
    }

    @JvmStatic
    @ZenMethod
    fun removeCrusherRecipe(@NotNull input: IItemStack) {
        MineTweakerAPI.apply(RemoveRecipeAction("WaterPower Crusher", RecipeManagers.crusher, MineTweakerMC.getItemStack(input)))
    }

    @JvmStatic
    @ZenMethod
    fun removeLatheRecipe(@NotNull input: IItemStack) {
        MineTweakerAPI.apply(RemoveRecipeAction("WaterPower Lathe", RecipeManagers.lathe, MineTweakerMC.getItemStack(input)))
    }

    @ZenMethod
    fun removeCutterRecipe(@NotNull input: IItemStack) {
        MineTweakerAPI.apply(RemoveRecipeAction("WaterPower Cutter", RecipeManagers.cutter, MineTweakerMC.getItemStack(input)))
    }

    @ZenMethod
    fun removeCompressorRecipe(@NotNull input: IItemStack) {
        MineTweakerAPI.apply(RemoveRecipeAction("WaterPower Compressor", RecipeManagers.compressor, MineTweakerMC.getItemStack(input)))
    }

    @ZenMethod
    fun removeCentrifugeRecipe(@NotNull input: IItemStack) {
        MineTweakerAPI.apply(RemoveRecipeAction("WaterPower Centrifuge", RecipeManagers.centrifuge, MineTweakerMC.getItemStack(input)))
    }

    @ZenMethod
    fun removeAdvancedCompressorRecipe(@NotNull input: IItemStack) {
        MineTweakerAPI.apply(RemoveRecipeAction("WaterPower Advanced Compressor", RecipeManagers.advCompressor, MineTweakerMC.getItemStack(input)))
    }

    @ZenMethod
    fun removeSawmillRecipe(@NotNull input: IItemStack, @NotNull output: IItemStack) {
        MineTweakerAPI.apply(RemoveRecipeAction("WaterPower Sawmill", RecipeManagers.sawmill, MineTweakerMC.getItemStack(input)))
    }

    private fun getMachineOutput(input: IItemStack, rm: IRecipeManager): IItemStack? {
        val output = rm.getOutput(MineTweakerMC.getItemStack(input), false)
        if (output == null || output.items.isEmpty())
            return null
        return MineTweakerMC.getIItemStack(output.items.first())
    }

    @JvmStatic
    @ZenMethod
    fun getCrusherOutput(input: IItemStack): IItemStack? {
        return getMachineOutput(input, RecipeManagers.crusher)
    }

    @JvmStatic
    @ZenMethod
    fun getAdvancedCompressorOutput(input: IItemStack): IItemStack? {
        return getMachineOutput(input, RecipeManagers.advCompressor)
    }

    @JvmStatic
    @ZenMethod
    fun getCentrifugeOutput(input: IItemStack): IItemStack? {
        return getMachineOutput(input, RecipeManagers.centrifuge)
    }

    @JvmStatic
    @ZenMethod
    fun getCompressorOutput(input: IItemStack): IItemStack? {
        return getMachineOutput(input, RecipeManagers.compressor)
    }

    @JvmStatic
    @ZenMethod
    fun getCutterOutput(input: IItemStack): IItemStack? {
        return getMachineOutput(input, RecipeManagers.cutter)
    }

    @JvmStatic
    @ZenMethod
    fun getLatheOutput(input: IItemStack): IItemStack? {
        return getMachineOutput(input, RecipeManagers.lathe)
    }

    @JvmStatic
    @ZenMethod
    fun getSawmillOutput(input: IItemStack): IItemStack? {
        return getMachineOutput(input, RecipeManagers.sawmill)
    }

}